frontend/src/pages/PipelineDetails.tsx

```diff
@@ -262,7 +274,20 @@ class PipelineDetails extends Page<{}, PipelineDetailsState> {
     if (fromRunId) {
       try {
         const runDetails = await Apis.runServiceApi.getRun(fromRunId);
-        templateYaml = RunUtils.getPipelineSpec(runDetails.run) || '';
+
+        // Convert the run's pipeline spec to YAML to be displayed as the pipeline's source.
+        try {
+          const pipelineSpec = JSON.parse(RunUtils.getPipelineSpec(runDetails.run) || '{}');
+          try {
+            templateString = JsYaml.safeDump(pipelineSpec);
+          } catch (err) {
+            await this.showPageError('Failed to convert pipeline spec to YAML.', err);
```

Comment:
Maybe we shouldn't expose that level of detail? The user likely doesn't know anything about both the JSON and YAML formats of the pipeline at this point. Maybe just "Failed to parse pipeline spec."? So maybe it can just be joined with the below catch.

Reply:
I think it's good to still keep the logger statements separate so we can debug it easier, but yeah, the user-facing messages don't need to be different here.